src/StateMachine.jl

```diff
@@ -240,5 +221,105 @@ function pretty(
+    # - A bound pattern to perform and then move on to the next state, or
+    # - An Expr to insert into the code when all else is exhausted
+    #   (which throws MatchFailure)
+    @_const action::Union{CasePartialResult, BoundPattern, Expr}
+
+    # The next code point(s):
+    # - Tuple{} if the action is a case which was matched or a MatchFailure
+    # - Tuple{DeduplicatedCodePoint} if the action was a fetch pattern. It designates
+    #   the code to perform after the fetch.
+    # - Tuple{DeduplicatedCodePoint, DeduplicatedCodePoint} if the action is a test.  These are the states
+    #   to go to if the result of the test is true ([1]) or false ([2]).
+    @_const next::Union{Tuple{}, Tuple{DeduplicatedCodePoint}, Tuple{DeduplicatedCodePoint, DeduplicatedCodePoint}}
+
+    @_const _cached_hash::UInt64
+    function DeduplicatedCodePoint(action, next)
+        action isa CasePartialResult && @assert action.pattern isa BoundTruePattern
+        new(nothing, action, next, hash((action, next)))
+    end
+end
+Base.hash(code::DeduplicatedCodePoint, h::UInt64) = hash(code._cached_hash, h)
+Base.hash(code::DeduplicatedCodePoint) = code._cached_hash
+function Base.:(==)(a::DeduplicatedCodePoint, b::DeduplicatedCodePoint)
+    a === b ||
+        a._cached_hash == b._cached_hash &&
+        isequal(a.action, b.action) &&
+        isequal(a.next, b.next)
+end
+function ensure_label!(code::DeduplicatedCodePoint, state::BinderState)
+    if code.label isa Nothing
+        code.label = gensym("label", state)
+    end
+end
+
+#
+# Deduplicate a code point, given the deduplications of the downstream code points.
+#
+function dedup(
+    dict::Dict{DeduplicatedCodePoint, DeduplicatedCodePoint},
+    code::CodePoint,
+    state::BinderState)
+    next = if code.next isa Tuple{}
+        code.next
+    elseif code.next isa Tuple{CodePoint}
+        (dedup(dict, code.next[1], state),)
+    elseif code.next isa Tuple{CodePoint, CodePoint}
+        t = dedup(dict, code.next[1], state)
+        f = dedup(dict, code.next[2], state)
+        # we might fall through to the true label, but we always jump to the false label
+        ensure_label!(f, state)
+        (t, f)
+    else
+        error("Unknown next type: $(code.next)")
+    end
+    key = DeduplicatedCodePoint(code.action, next)
+    result = get!(dict, key, key)
+    if result !== key
+        # The state already existed, so it must have had two predecessors
+        # We will need a label for one of them to use in the generated code
+        ensure_label!(result, state)
+    end
+    result
+end
+
+#
+# Deduplicate the state machine by collapsing behaviorally identical states.
+#
+function deduplicate_state_machine(entry::CodePoint, state::BinderState)
+    dedup_map = Dict{DeduplicatedCodePoint, DeduplicatedCodePoint}()
+    result = Vector{DeduplicatedCodePoint}()
+    top_down_states = reachable_states(entry)
+    bottom_up_states = reverse(top_down_states)
+    for e in bottom_up_states
```

Comment:
```suggestion
    for e in Iterators.reverse(top_down_states)
```
to avoid the copy - will need the import